gestures.js

```diff
@@ -1,4 +1,5 @@
 const Module = imports.misc.extensionUtils.getCurrentExtension().imports.module;
+const Settings = Module.Settings();
```

Comment:
Does this then introduce potential dependency cycles again?

Should we maybe add a comment to `settings.js` that we should not import other modules there?

Reply:
Also why was it needed to put this on the top level again?

Reply:
ah, settings is a bit different - we need more than the exported functions - we need the exported prefs var and the gsettings connect functionality (which then basically allow those modules to also become "listeners" to gsetting pref change signals).

Remember imports in gnome extensions basically just copies the exported functions & vars at the top level (in gnome extensions, top level functions and vars layer are exported).  You'll notice the modules that need this are the same ones that need to use `gsettings.connect(..., ....);` functions (e.g. they need to be notified when user prefs are changed).

If you don't do this, then when a user changes a setting (like say `show-window-position-bar`) `topbar.js` won't be informed of this change and do nothing (until PaperWM is restarted).  Same with the other modules.

All other `Module` usages basically just need access to  (independent) functions - hence we don't need to do a full import of other modules at the top (which can produce some side-effects, especially if that module imports other modules).

Anyways, this one tripped me up for a bit since I went to change a setting in prefs and PaperWM did nothing (didn't pick it up until it was restarted).

Reply:
Hope that makes sense.  So, basically, if you just want to use an (independent) function, then `Modules.Blad().function(...)` if sufficient and clean (it's basically just like copy/pasting just that function in the module).  However, if you need more than that - e.g. if an implementation depends on creation of objects, gsetting connections, etc. (basicaly more than independent functions) - then it may need to be fully imported at the top.

Reply:
> Does this then introduce potential dependency cycles again?
> 
> Should we maybe add a comment to `settings.js` that we should not import other modules there?

No, not here, but just because `settings.js` (and also `utils.js`) are a little special and a `pure` modules in the sense that they don't import other modules (or only rely on Module referencing) - see comments above.

Reply:
Ok. If it works, it's fine.

Reply:
If you do a function call like this in global scope it will be rejected on ego.

btw, importing modules inside a function is not a good practice and if you do that, you will have a hard time porting it to GNOME Shell 45.

Reply:
> If you do a function call like this in global scope it will be rejected on ego.
> 
> btw, importing modules inside a function is not a good practice and if you do that, you will have a hard time porting it to GNOME Shell 45.

@jrahmatzadeh!  fair enough - will update that.

 Given we've done a boatload of refactoring and changes, and better separating functions into more appropriate modules, we might be able to use imports for those modules (and clean up what gets exported).

Is there something in gnome 45 that impacts doing `imports` within functions?

Reply:
Yes, 45 will move to ESM, so import should be at the top of the file. You should wait for the [port guide](https://gjs.guide/extensions/#upgrading) though. We're going to discuss about the imports on GUADEC. _Maybe_ we can come up with a solution for backward compatibility.

Reply:
Thanks so much for the heads-up @jrahmatzadeh!  Yep, good call then.  

Although I did think the `Module.js` approach was cool it at least it made it easier to figure out all the circular deps that were going on...

I've now removed it and put all imports at the top (and took the opportunity to clean up a bunch of stuff).

Hopefully things are looking much better for a new submission to ego.

Reply:
@Lythenas, check out the latest here - I removed `module.js` (I thought it was cool while it lasted...).  Hoping to merge this one info `develop`, and `release` soon!

Reply:
Refactored modules to be more appropriate for whether they're needed in gnome-shell or prefs.js.  Main issues was an inner dependency in settings.js on utils.js (and utils.js imported a bunch of stuff not available to `prefs.js`.  To work around that we were not importing utils.js at the top... but doing it in enable.  Anyways, this is now solved and all modules are imported at the top now.  It's also much cleaner.